### ISSUE TITLE
[2D] Simplify parallax and sorting workflow

### ISSUE THREAD

**Issue by julhe** [godotengine/godot-proposals#4963]
### Describe the project you are working on

A 2D side scroller game.

### Describe the problem or limitation you are having in your project

We have many sprites that need to be aranged on slightly different depths. Creating a ParallaxLayer for each sprite creates a lot of clutter in the scene. It also tents to be error prone as one might forget the put the many ParallaxLayers in the correct sorting order.

Furthermore, we would like to do a vertigo effect in a few instances (changing the z-offset and fov). Coming from Unity both of these were easy to pull off in a 3D environment, but would be a huge stretch with the current workflow in Godot.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Move the entire workflow closer to how sorting and parallax would work in 3D.

Having a single value (Z-Position) that affects both sorting and parallax would make the workflow more intuitive, faster to work on and less error prone.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Deprecated ParallaxLayers and Z Index in favour of a single Node2D parameter called Z-position, which affects both sorting and parallax behaviour. Camera2D will additionally have something as a field-of-view or depth-scale value that affects how much the Z-position contributes to the actual sprite size and an additionally base Z offset. 

As a start, the existing Z Index infrastructure already affects the sorting and might be easily repurposed to also affect the parallax scale.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Sorting and depth are naturally tied together, so not much would be lost in my opinion. For some special effects these two might need to be separated however. That could be done with a second value called something like "Sorting Offset" which would cover this use-case. 

### Is there a reason why this should be core and not an add-on in the asset library?

The ways parallax and sorting are done in Godot are hardly changeable from an add-on, as far as I could tell.

**Comment by YuriSizov** [comment]
Unlike Unity, Godot's 2D is actual 2D, and not a projection of 3D, so adding Z-position doesn't make sense. It should still be Z-index, but the idea to have in-camera effects based on the Z-index sounds interesting to me. That said, ParalaxLayer creates a fake 3D effect. If you want to work with 3D and fake 2D instead, you can use Godot's 3D side and Sprite3D nodes, for example.

**Comment by julhe** [comment]
Maybe the problem is more clear if I show how our test scene looks:
![grafik](https://user-images.githubusercontent.com/18037091/180971960-254320e8-bdb6-4a48-8766-c018b9f97b32.png)
Changing anything here without breaking the sorting in subtle ways gets harder the more nodes are added.

So, I just want to change how scale and sorting come together (no "3D stuff"):
- Every Node2D has a "follow viewport scale", not just the ParallaxLayer. No need to add a parallax layer for every different object. 
- Because "follow viewport scale" affects parallax, it should also affect the sorting. Now sorting is automatic and the scene remains easy to change. 
- A value in the Camera2D determents how strong that parallax scale should be. If no parallax is wanted, It can be disabled here.

Also, users who don't need parallax could simply ignore the Z-Depth/Position/Value at all and remain using the sorting implied by the hierarchy.

On the 3D-Side: Switching to 3D would mean additional work and honestly feels weird, because we just have a workflow problem and don't intend to use any "real" 3D features at all.

**Comment by YuriSizov** [comment]
> Also, users who don't need parallax could simply ignore the Z-Depth/Position/Value at all and remain using the sorting implied by the hierarchy.

Z-sorting is very much useful on its own, without any parallax. So it's an existing feature that resolves specific functionality, it wouldn't be appropriate to rework it to solve something else.

> On the 3D-Side: Switching to 3D would mean additional work and honestly feels weird, because we just have a workflow problem and don't intend to use any "real" 3D features at all.

All I'm saying is that if you want a workflow similar to Unity you need to use 3D in Godot. It would be easier than trying to make fake 3D of parallax layers to do that work for you. You're also using ParallaxLayer without ParallaxBackground, it seems? Maybe all you need is CanvasLayers, which also have their own sorting index.

**Comment by julhe** [comment]
> Z-sorting is very much useful on its own, without any parallax. So it's an existing feature that resolves specific functionality, it wouldn't be appropriate to rework it to solve something else.

I don't think thats the case, as parallax and sorting are tied together. There is no common use-case were you want something that is futher away to be rendered all up front, so why not link it? 

> All I'm saying is that if you want a workflow similar to Unity you need to use 3D in Godot. It would be easier than trying to make fake 3D of parallax layers to do that work for you. 

I'm still hesitant to switch to 3D, after all 2D content support is basic (no Drag n' Drop, no sprite meshes) and theres no support for Spine. 

> You're also using ParallaxLayer without ParallaxBackground, it seems? Maybe all you need is CanvasLayers, which also have their own sorting index.

That doesn't solve the problem. I would still need to create a new layer for each sprite.
___
I think it would be still ok if I had at least the abillity to do parallax scale per node. With that, I could work my way arround it.

**Comment by YuriSizov** [comment]
> I don't think thats the case, as parallax and sorting are tied together. There is no common use-case were you want something that is futher away to be rendered all up front, so why not link it?

Yes, the parallax effect is related to depth sorting. But what I'm saying is that existing z-index/z-sorting is not exclusive to parallax, it's not tied to it. If I understand correctly your proposal, you want them to be deeply mutually related, but that would conflict with existing usages of z-index.

The built-in parallax effect in Godot is based on canvas layers, which is why `ParallaxBackground` must be used. Your example is missing it for some reason, so you are already misusing the feature. `ParallaxBackground`, being a `CanvasLayer`, already has a sorting parameter `layer` that is connected to its visual depth and drawing order. `CanvasLayer` also has scale, by the way, have you looked into that?

> That doesn't solve the problem. I would still need to create a new layer for each sprite.

You would need to create a new layer for each _plane_, not for each sprite. I'm not sure what specific artistic limitations your project has, but I can't imagine there being thousands of individual layers, one for each sprite, each with its own scale and its own motion offset. Maybe you'd have a couple of dozen unique layers, which `CanvasLayer`s/`ParallaxBackground`s are more than capable to handle. If you indeed need thousands of micro adjustments for individual sprites, that should probably be done within canvas layers which define the major planes.

**Comment by julhe** [comment]
It turned out that the changes for 3D are doable and we're switching to it. Not sure if to close this proposal, as I still think that the 2D workflow has some issues (see my answer below)?
___
> Yes, the parallax effect is related to depth sorting. But what I'm saying is that existing z-index/z-sorting is not exclusive to parallax, it's not tied to it. If I understand correctly your proposal, you want them to be deeply mutually related, but that would conflict with existing usages of z-index.

No, the relation would be optional. Likely the Camera2D would controll how much the Z-index/Offset contributes to the final parallax scale.

> The built-in parallax effect in Godot is based on canvas layers, which is why `ParallaxBackground` must be used. Your example is missing it for some reason, so you are already misusing the feature. `ParallaxBackground`, being a `CanvasLayer`, already has a sorting parameter `layer` that is connected to its visual depth and drawing order. `CanvasLayer` also has scale, by the way, have you looked into that?

Yes, thats what I'm using currently. With visual depth im refering to the parallax, sorry if that's confusing.

> You would need to create a new layer for each _plane_, not for each sprite. I'm not sure what specific artistic limitations your project has, but I can't imagine there being thousands of individual layers, one for each sprite, each with its own scale and its own motion offset. 

As for reference, the Ori games are close of what we envision, at least form the use of parallax, see https://youtu.be/fXUrR6EiEcY?t=4414 for example. (Espeically the detail vegetation has many different parallax levels, which would be lots of layers.)

> Maybe you'd have a couple of dozen unique layers, which `CanvasLayer`s/`ParallaxBackground`s are more than capable to handle. If you indeed need thousands of micro adjustments for individual sprites, that should probably be done within canvas layers which define the major planes.

Not sure if I correctly understood what you suggested, but I've generated 30 Layers ahead and made a scene. That actually works, but there are still problems:

- No way to create prefabs with multiple parallax levels, since they span across the major layers. (Could be possible to make sub parallax layers, but I coudn't figure out a way to make them inherint the scale of thier parrent -> allow them to be moved to different depths. )
- Hard to anticipate which layers to create in advance: I sometimes want to nudge a sprite on top of another one (parallax wise). That might be nit picking, but going into the process of creating a new layer each time breaks the flow for me.
